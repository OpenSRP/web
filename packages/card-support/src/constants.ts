// OPENSRP API endpoints
export const OPENSRP_URL_USER_ASSIGNMENT = '/organization/user-assignment';
export const OPENSRP_URL_CLIENT_SEARCH = '/client/search';
