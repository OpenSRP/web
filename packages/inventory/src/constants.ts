// opensrp api strings
export const OPENSRP_API_BASE_URL = 'https://mg-eusm-staging.smartregister.org/opensrp/rest/';

// router routes
export const INVENTORY_SERVICE_POINT_LIST_VIEW = '/inventory/list';
export const INVENTORY_SERVICE_POINT_PROFILE_VIEW = '/inventory/list';

// other constants
/** namespace for the keys attached to the columns */
export const TableColumnsNamespace = 'inventory';

// magic strings
export const SEARCH_QUERY_PARAM = 'querySearch';
export const TABLE_PAGE_SIZE = 5;
export const TABLE_PAGE_SIZE_OPTIONS = ['5', '10', '20', '50', '100'];

export const tablePaginationOptions = {
  showQuickJumper: true,
  showSizeChanger: true,
  defaultPageSize: TABLE_PAGE_SIZE,
  pageSizeOptions: TABLE_PAGE_SIZE_OPTIONS,
};

//endpoints
export const GET_INVENTORY_BY_SERVICE_POINT = 'stockresource/servicePointId/';

//payload
export const inventory1 = {
  stockId: '5dfcaecd-430a-4284-972f-9acf87775a9b1',
  productName: 'Scale1',
  unicefSection: 'Health1',
  quantity: 4,
  deliveryDate: '2020-02-01',
  accountabilityEndDate: '2020-10-10',
  donor: 'ADB',
  servicePointId: '903971',
  poNumber: 1111,
  serialNumber: '1234serial2',
  providerId: 'providerId',
};

export const inventory2 = {
  stockId: '5dfcaecd-430a-4284-972f-9acf87775a9b2',
  productName: 'Scale2',
  unicefSection: 'Health2',
  quantity: 4,
  deliveryDate: '2020-02-01',
  accountabilityEndDate: '2020-10-10',
  donor: 'ADB',
  servicePointId: '903972',
  poNumber: 1112,
  serialNumber: '1234serial2',
  providerId: 'providerId',
};
