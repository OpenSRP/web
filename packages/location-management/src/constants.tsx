// URLs
export const URL_ALL_LOCATION_TAGS = '/location-tag';
export const URL_DELETE_LOCATION_TAGS = '/location-tag/delete';
export const KEYCLOAK_API_BASE_URL ='https://opensrp-stage.smartregister.org/opensrp/rest';
