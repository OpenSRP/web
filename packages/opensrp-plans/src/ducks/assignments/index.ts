/** Assignments redux module */
import { Dictionary } from '@onaio/utils';
import { get, uniqWith } from 'lodash';
import moment from 'moment';
import { Store } from 'redux';
import { AnyAction } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'assignments';

/** the raw assignment object as received from openSRP API */
export interface RawAssignment {
  jurisdictionId: string;
  organizationId: string;
  planId: string;
  fromDate: number;
  toDate: number;
}

/** interface for a Assignment object */
export interface Assignment {
  jurisdiction: string;
  organization: string;
  plan: string;
  fromDate: string;
  toDate: string;
}

// action interfaces

/** action type for action that adds Assignments to store */
export const ASSIGNMENTS_FETCHED = 'src/store/ducks/assignments/reducer/ASSIGNMENTS_FETCHED';
/** action type for REMOVE_TEAMS action */
export const REMOVE_ASSIGNMENTS = 'src/store/ducks/assignments/reducer/REMOVE_ASSIGNMENTS';

/** interface for Assignments fetched action */
export interface FetchAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  overwrite: boolean;
  type: typeof ASSIGNMENTS_FETCHED;
}

/** interface for action that removes assignments from store */
export interface RemoveAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof REMOVE_ASSIGNMENTS;
}

/** single type for all action types */
type AssignmentActionTypes = FetchAssignmentsAction | RemoveAssignmentsAction | AnyAction;

/** interface describing the assignmentsByPlanId Store */
interface AssignmentsByPlanId {
  [key: string]: Assignment[];
}
/** interface for Assignments state in store */
interface AssignmentsStoreState {
  assignmentsByPlanId: AssignmentsByPlanId | {};
}

// immutable assignments state in dux
export type ImmutableAssignmentsStoreState = AssignmentsStoreState &
  SeamlessImmutable.ImmutableObject<AssignmentsStoreState>;

/** initial state for Assignments records in store */
const initialAssignmentStoreState: ImmutableAssignmentsStoreState = SeamlessImmutable({
  assignmentsByPlanId: {},
});

/**
 * the Assignment reducer function
 *
 * @param {object} state - the store
 * @param {object} action - the action types
 * @returns {object} - the state
 */
export default function reducer(
  state = initialAssignmentStoreState,
  action: AssignmentActionTypes
) {
  switch (action.type) {
    case ASSIGNMENTS_FETCHED:
      if (!action.overwrite) {
        // so what we want to do is to ensure all action.assignmentsByPlanId arrays of the
        // same plan are merged.  But while merging we want to that the array elements are
        // unique based on fromDate, jurisdiction and organization
        const currentState = state.assignmentsByPlanId.asMutable();
        // loop through each plan in the action object
        for (const [planId, assignments] of Object.entries(action.assignmentsByPlanId)) {
          // check if the plan is already in current state
          if (planId in currentState) {
            // merge the two assignment arrays
            const allAssignments = (assignments as Assignment[]).concat(
              (currentState as AssignmentsByPlanId)[planId]
            );
            // sort by whether an element was NOT in the current state i.e. elements NOT in
            // current state will be ordered first.  We do this because we want to remove elements
            // that were in current state.  We assume that they are being overwritten
            allAssignments.sort((_, b) => ((assignments as Assignment[]).includes(b) ? 1 : -1));
            // remove the elements in the current state.  This works because we hard already ordered
            // the elements.  uniqWith keeps the first found item
            const filteredAssignments = (action.assignmentsByPlanId[planId] = uniqWith(
              allAssignments,
              (a, b) =>
                `${a.fromDate}${a.jurisdiction}${a.organization}` ===
                `${b.fromDate}${b.jurisdiction}${b.organization}`
            ));
            // finally, save this to the action object
            action.assignmentsByPlanId[planId] = filteredAssignments;
          }
        }
      }

      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: {
          ...state.assignmentsByPlanId,
          ...action.assignmentsByPlanId,
        },
      });
    case REMOVE_ASSIGNMENTS:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: action.assignmentsByPlanId,
      });
    default:
      return state;
  }
}

/** action to remove assignments form store */
export const removeAssignmentsAction: RemoveAssignmentsAction = {
  assignmentsByPlanId: {},
  type: REMOVE_ASSIGNMENTS,
};

// action creators

/** creates action to add fetched assignments to store
 *
 * @param {object []} assignmentsList - array of assignments to be added to store
 * @param {boolean} overwrite - whether to overwrite assignments
 * @returns {object} - action with assignments payload that is added to store
 */
export const fetchAssignments = (
  assignmentsList: Assignment[],
  overwrite = false
): FetchAssignmentsAction => {
  const assignmentsByPlanId: Dictionary<Assignment[]> = {};
  assignmentsList.forEach((assignment) => {
    const planId = assignment.plan;
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!assignmentsByPlanId[planId]) {
      assignmentsByPlanId[planId] = [];
    }
    assignmentsByPlanId[assignment.plan].push(assignment);
  });
  return {
    assignmentsByPlanId,
    overwrite,
    type: ASSIGNMENTS_FETCHED,
  };
};

/** creates action to reset plan
 *
 * @param {object} assignmentsByPlanId object with updated assignment arrays, keyed by planId
 * @returns {object} -
 */
export const resetPlanAssignments = (
  assignmentsByPlanId: AssignmentsByPlanId = {}
): FetchAssignmentsAction => {
  return {
    assignmentsByPlanId,
    overwrite: true,
    type: ASSIGNMENTS_FETCHED,
  };
};

// selectors

/** get assignments as an object where their ids are the keys and the objects
 * the values
 *
 * @param {object} state - Portion of the store
 *
 * @returns {object} -
 */
export function getAssignmentsByPlanId(state: Partial<Store>): { [key: string]: Assignment[] } {
  return (state as Dictionary)[reducerName].assignmentsByPlanId;
}

/** Get all assignments as an array
 *
 * @param {object} state - Part of the redux store
 * @param {string} planId - The id of the plan of which to return associated Assignments
 *
 * @returns {object} - all assignments by planId in store as an array
 */
export function getAssignmentsArrayByPlanId(state: Partial<Store>, planId: string): Assignment[] {
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  const assignments = get(getAssignmentsByPlanId(state), planId) ?? [];
  return assignments.filter((obj) => moment(obj.toDate) >= moment());
}
