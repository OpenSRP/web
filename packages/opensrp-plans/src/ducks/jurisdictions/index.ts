import { Dictionary } from '@onaio/utils/dist/types/types';
import {
  fetchActionCreatorFactory,
  reducerFactory,
  removeActionCreatorFactory,
} from '@opensrp/reducer-factory';
import { Geometry } from 'geojson';
import { get, values } from 'lodash';
import { Store } from 'redux';
import { createSelector } from 'reselect';

export interface LocationTag {
  id: number;
  name: string;
}

/** The shape of a jurisdiction received from the OpenSRP API */
export interface Jurisdiction {
  id: string;
  geometry?: Geometry;
  properties: {
    code?: string;
    geographicLevel: number;
    name: string;
    parentId?: string;
    status: string;
    version: string | number;
    OpenMRS_Id?: string;
    externalId?: string;
    username?: string;
  };
  locationTags: LocationTag[];
  serverVersion: number;
  type: Readonly<'Feature'>;
}

/** the reducer name */
export const jurisdictionReducerName = 'Jurisdiction';

/** Jurisdiction Reducer */
export const JurisdictionsReducer = reducerFactory<Jurisdiction>(jurisdictionReducerName);

// actions
/** actionCreator returns action to to add Item records to store */
export const fetchJurisdictions = fetchActionCreatorFactory<Jurisdiction>(
  jurisdictionReducerName,
  'id'
);
export const removeJurisdictions = removeActionCreatorFactory(jurisdictionReducerName);

// selectors
/** prop filters to customize selector queries */
export interface Filters {
  jurisdictionId?: string /** jurisdiction id */;
  geoLevel?: number /** return jurisdictions with this geo level */;
}

/**
 * retrieve the jurisdictionId value
 *
 * @param {object} _ - the store
 * @param {object} props -  the filterProps
 * @returns {string} -
 */
export const getJurisdictionId = (_: Partial<Store>, props: Filters) => props.jurisdictionId;

/**
 * retrieve the geoLevel value
 *
 * @param {object} _ - the store
 * @param {object} props -  the filterProps
 * @returns {boolean} -
 */
export const getGeoLevel = (_: Partial<Store>, props: Filters) => props.geoLevel;

/** gets all jurisdictions keyed by id
 *
 * @param {object} state - the store
 * @param {object} _ -  the filterProps
 * @returns {Dictionary<object>} -
 */
export const getJurisdictionsById = (state: Partial<Store>, _: Filters): Dictionary<Jurisdiction> =>
  (state as Dictionary)[jurisdictionReducerName].objectsById;

/** retrieve the Jurisdiction using an id
 *
 * @returns {object} -
 */
export const getJurisdictionById = () =>
  createSelector(
    [getJurisdictionsById, getJurisdictionId],
    (jurisdictionsById, jurisdictionId): Jurisdiction | null => {
      if (!jurisdictionId) {
        return null;
      }
      return get(jurisdictionsById, jurisdictionId, null);
    }
  );

/**
 * retrieve jurisdictions as an array filtered by geoLevel
 *
 * @returns {object} -
 */
export const getJursByGeoLevel = () =>
  createSelector(
    [getJurisdictionsById, getGeoLevel],
    (jurisdictionsById, geoLevel): Jurisdiction[] => {
      const result = values(jurisdictionsById);
      if (geoLevel) {
        return result.filter((item) => item.properties.geographicLevel === geoLevel);
      }
      return result;
    }
  );
