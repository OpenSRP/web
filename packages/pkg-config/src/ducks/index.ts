import { createSelector, Store, createSlice, PayloadAction } from '@reduxjs/toolkit';
export const sliceName = 'configs';

export interface ConfigState {
  [key: string]: string;
}

export const defaultConfigState: ConfigState = {};

export const configsSlice = createSlice({
  name: sliceName,
  initialState: defaultConfigState,
  reducers: {
    addConfigs(state, action: PayloadAction<ConfigState>) {
      return {
        ...state,
        ...action.payload,
      };
    },
  },
});

const getConfigsState = (state: Partial<Store>) => {
  return (state as any)[sliceName];
};

export const getConfigsFactory = () => createSelector(getConfigsState, (state) => state);

export const { addConfigs } = configsSlice.actions;
export const { reducer } = configsSlice;
