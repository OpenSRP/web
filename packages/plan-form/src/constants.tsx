export const A1 = 'A1';
export const A2 = 'A2';
export const B1 = 'B1';
export const B2 = 'B2';
export const LOW_PRIORITY = 'low-priority';
export const MEDIUM_PRIORITY = 'medium-priority';
export const HIGH_PRIORITIY = 'high-priority';
export const TRUE = 'True';
export const FALSE = 'False';
export const DISABLED = 'Disabled';
export const INTERNAL = 'internal';
export const IGNORE = 'ignore';
export const INTERVENTION_TYPE_CODE = 'interventionType';
export const FI_STATUS_CODE = 'fiStatus';
export const FI_REASON_CODE = 'fiReason';
export const OPENSRP_EVENT_ID_CODE = 'opensrpEventId';
export const CASE_NUMBER_CODE = 'caseNum';
export const TASK_GENERATION_STATUS_CODE = 'taskGenerationStatus';
export const TEAM_ASSIGNMENT_STATUS_CODE = 'teamAssignmentStatus';
export const CASE_TRIGGERED = 'Case Triggered';
export const ROUTINE = 'Routine';
export const INVESTIGATION = 'Investigation';
export const DAYS = 'days';
export const CONDITION = 'condition';
export const TRIGGER = 'trigger';

// plan activity codes
export const CASE_CONFIRMATION_ACTIVITY_CODE = 'caseConfirmation';
export const FAMILY_REGISTRATION_ACTIVITY_CODE = 'familyRegistration';
// opensrp api strings
export const PLAN_LIST_URL = '/plans/list';
export const OPENSRP_API_BASE_URL = '';
export const OPENSRP_PLAN_ENDPOINT = 'plans';