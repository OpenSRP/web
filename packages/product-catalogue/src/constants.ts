// opensrp api strings
export const OPENSRP_PRODUCT_CATALOGUE = 'rest/product-catalogue';

// router routes
export const CATALOGUE_LIST_VIEW_URL = '/product-catalogue';
export const CATALOGUE_EDIT_VIEW_URL = '/product-catalogue/edit';
export const CATALOGUE_CREATE_VIEW_URL = 'product-catalogue/new';
export const HOME_URL = '/';

// other constants
/** namespace for the keys attached to the columns */
export const TableColumnsNamespace = 'product-catalogue';

/** route params for product-catalogue pages */
export interface RouteParams {
  productId?: string;
}
