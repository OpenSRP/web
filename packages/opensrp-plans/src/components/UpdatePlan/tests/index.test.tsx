import React from 'react';
import { ConnectedEditPlanView } from '..';
import { store } from '@opensrp/store';
import { createBrowserHistory } from 'history';
import { Router } from 'react-router';
import { Provider } from 'react-redux';
import { eusmPlans } from '../../../ducks/tests/fixtures';
import { PLANS_LIST_VIEW_URL } from '../../../constants';
import { mount } from 'enzyme';
import { Helmet } from 'react-helmet';
import { PlanForm } from '@opensrp/plan-form';
import { act } from 'react-dom/test-utils';
import { removePlanDefinitions } from '../../../ducks';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('jest-fetch-mock');

const mission1 = eusmPlans[0];

const history = createBrowserHistory();

describe('CreateEditProduct Page', () => {
  beforeEach(() => {
    store.dispatch(removePlanDefinitions());
  });

  afterEach(() => {
    fetch.resetMocks();
  });

  it('renders correctly with store(for editing plans)', async () => {
    fetch.mockResponse(JSON.stringify([mission1]));

    const props = {
      history,
      location: {
        hash: '',
        pathname: `${PLANS_LIST_VIEW_URL}/${mission1.identifier}`,
        search: '',
        state: {},
      },
      match: {
        isExact: true,
        params: { planId: mission1.identifier },
        path: `${PLANS_LIST_VIEW_URL}/:planId`,
        url: `${PLANS_LIST_VIEW_URL}/${mission1.identifier}`,
      },
    };

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedEditPlanView {...props}></ConnectedEditPlanView>
        </Router>
      </Provider>
    );

    // loading
    expect(wrapper.text()).toMatchInlineSnapshot(
      `"Loading...Fetching planPlease wait, as we fetch the plan."`
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('full text snapshot');

    // check if page title is correct
    const helmet = Helmet.peek();
    expect(helmet.title).toEqual('Edit > Scale');

    // check if form is rendered on the page
    expect(wrapper.find('form')).toHaveLength(1);

    expect(wrapper.find(PlanForm).props()).toMatchSnapshot('edit form props');
  });

  it('shows broken page', async () => {
    const props = {
      history,
      location: {
        hash: '',
        pathname: `${PLANS_LIST_VIEW_URL}/${mission1.identifier}`,
        search: '',
        state: {},
      },
      match: {
        isExact: true,
        params: { planId: mission1.identifier },
        path: `${PLANS_LIST_VIEW_URL}/:planId`,
        url: `${PLANS_LIST_VIEW_URL}/${mission1.identifier}`,
      },
    };

    fetch.mockReject(new Error('Could not pull data'));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedEditPlanView {...props}></ConnectedEditPlanView>
        </Router>
      </Provider>
    );

    // show loading screen
    expect(wrapper.text()).toMatchSnapshot('full page text');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // should be in error page
    expect(wrapper.text()).toMatchInlineSnapshot(`"ErrorCould not pull dataGo BackBack Home"`);
  });

  it('handles missing plan', async () => {
    const props = {
      history,
      location: {
        hash: '',
        pathname: `${PLANS_LIST_VIEW_URL}/${mission1.identifier}`,
        search: '',
        state: {},
      },
      match: {
        isExact: true,
        params: { planId: mission1.identifier },
        path: `${PLANS_LIST_VIEW_URL}/:planId`,
        url: `${PLANS_LIST_VIEW_URL}/${mission1.identifier}`,
      },
    };

    fetch.mockResponse(JSON.stringify([mission1]));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedEditPlanView {...props}></ConnectedEditPlanView>
        </Router>
      </Provider>
    );

    // should be in loading screen
    expect(wrapper.text()).toMatchInlineSnapshot(
      `"Loading...Fetching planPlease wait, as we fetch the plan."`
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    /** resource404 info page */
    expect(wrapper.text()).toMatchInlineSnapshot(
      `"404Sorry, the resource you requested for, does not existGo BackBack Home"`
    );
  });
});
