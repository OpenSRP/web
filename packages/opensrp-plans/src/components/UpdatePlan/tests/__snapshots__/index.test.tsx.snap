// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEditProduct Page shows broken page: full page text 1`] = `"Loading...Fetching planPlease wait, as we fetch the plan."`;
